### PR TITLE
*: add constraint for the length of the index prefix

### PR DESCRIPTION
**Pull request by tiancaiamao**
See https://github.com/pingcap/tidb/issues/1429

it seems that MySQL use 3072 for length limit if the column doesn't specify blob length or specify a number larger than 255, otherwise the constraint length is 255.

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -52,6 +52,10 @@ func buildIndexInfo(tblInfo *model.TableInfo, unique bool, indexName model.CIStr
 			return nil, errors.Trace(errIncorrectPrefixKey)
 		}
 
+		if ic.Length > 255 {
```
Comment (shenli) [inline ddl/index.go]: "For example, a prefix can be up to 767 bytes long for InnoDB tables or 3072 bytes if the innodb_large_prefix option is enabled."
See: http://dev.mysql.com/doc/refman/5.7/en/create-index.html

We can use the constraint of InnoDB. And it is better to use a const.

File: ddl/index.go [hunk 2]
```diff
@@ -30,6 +30,8 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
+const prefixLength = 767
```
Comment (shenli) [inline ddl/index.go]: maxPrefixLength is better.

**Comment by shenli** [comment]
LGTM

File: session_test.go [hunk 3]
```diff
@@ -2147,6 +2147,10 @@ func (s *testSessionSuite) TestSpecifyIndexPrefixLength(c *C) {
 	// ERROR 1170 (42000): BLOB/TEXT column 'c2' used in key specification without a key length
 	c.Assert(err, NotNil)
 
+	_, err = exec(c, se, "create index idx_c1 on t (c2(555555));")
+	// ERROR 1071 (42000): Specified key was too long; max key length is 3072 bytes
```
Comment (coocood) [inline session_test.go]: This comment is inconsistent with the 'maxPrefixLength' value 767.

**Comment by coocood** [comment]
LGTM

**Pull request merged**